File: internal/executor/goexec/mcp_elicitation.go
package goexec

import (
	"encoding/json"
	"net"
	"net/url"
	"os"
	"path/filepath"
	"sort"
	"strconv"
	"strings"

	"everything-go/internal/backend"
)

const (
	browserOriginTool = "access_browser_origin"
	elicitationChoice = "decision"
)

type browserOriginPolicy struct {
	mode           string
	allowedOrigins []string
}

func ensureBrowserElicitationRouting(codexHome string) (bool, error) {
	manage := strings.ToLower(strings.TrimSpace(os.Getenv("BRIDGE_BROWSER_MANAGE_AUTO_REVIEW")))
	switch manage {
	case "0", "false", "no", "off":
		return false, nil
	}
	if codexHome == "" {
		codexHome = os.Getenv("CODEX_HOME")
	}
	if codexHome == "" {
		home, err := os.UserHomeDir()
		if err != nil {
			return false, err
		}
		codexHome = filepath.Join(home, ".codex")
	}
	dir := filepath.Join(codexHome, "browser")
	path := filepath.Join(dir, "config.toml")
	original, err := os.ReadFile(path)
	if err != nil && !os.IsNotExist(err) {
		return false, err
	}
	lines := strings.Split(string(original), "\n")
	found := false
	for index, line := range lines {
		trimmed := strings.TrimSpace(line)
		if strings.HasPrefix(trimmed, "disable_auto_review") {
			if eq := strings.Index(trimmed, "="); eq >= 0 && strings.TrimSpace(trimmed[:eq]) == "disable_auto_review" {
				lines[index] = "disable_auto_review = true"
				found = true
				break
			}
		}
	}
	updated := strings.Join(lines, "\n")
	if !found {
		updated = "disable_auto_review = true\n" + string(original)
	}
	if updated == string(original) {
		return false, nil
	}
	if err := os.MkdirAll(dir, 0o700); err != nil {
		return false, err
	}
	temporary, err := os.CreateTemp(dir, ".config.toml.*")
	if err != nil {
		return false, err
	}
	temporaryPath := temporary.Name()
	defer os.Remove(temporaryPath)
	if _, err = temporary.WriteString(updated); err == nil {
		err = temporary.Sync()
	}
	if closeErr := temporary.Close(); err == nil {
		err = closeErr
	}
	if err != nil {
		return false, err
	}
	mode := os.FileMode(0o600)
	if info, statErr := os.Stat(path); statErr == nil {
		mode = info.Mode().Perm()
	}
	if err := os.Chmod(temporaryPath, mode); err != nil {
		return false, err
	}
	if err := os.Rename(temporaryPath, path); err != nil {
		return false, err
	}
	return true, nil
}

func browserOriginPolicyFromEnv() browserOriginPolicy {
	mode := strings.ToLower(strings.TrimSpace(os.Getenv("BRIDGE_BROWSER_ORIGIN_MODE")))
	switch mode {
	case "ask", "allowlist", "allow_all", "deny":
	default:
		mode = "ask"
	}
	var allowed []string
	for _, item := range strings.Split(os.Getenv("BRIDGE_BROWSER_ALLOWED_ORIGINS"), ",") {
		if item = strings.TrimSpace(item); item != "" {
			allowed = append(allowed, item)
		}
	}
	return browserOriginPolicy{mode: mode, allowedOrigins: allowed}
}

func decodeObject(raw json.RawMessage) map[string]any {
	var out map[string]any
	if json.Unmarshal(raw, &out) != nil {
		return nil
	}
	return out
}

func recursiveValue(value any, key string, depth int) (any, bool) {
	if depth < 0 {
		return nil, false
	}
	switch item := value.(type) {
	case map[string]any:
		if found, ok := item[key]; ok {
			return found, true
		}
		for _, child := range item {
			if found, ok := recursiveValue(child, key, depth-1); ok {
				return found, true
			}
		}
	case []any:
		for _, child := range item {
			if found, ok := recursiveValue(child, key, depth-1); ok {
				return found, true
			}
		}
	}
	return nil, false
}

func recursiveString(value any, key string) string {
	found, _ := recursiveValue(value, key, 8)
	return strings.TrimSpace(codexAnyString(found))
}

func browserOriginRequest(params map[string]any) (tool, origin string, ok bool) {
	if recursiveString(params, "connector_id") != "browser-use" {
		return "", "", false
	}
	tool = recursiveString(params, "tool_name")
	origin = recursiveString(params, "origin")
	return tool, origin, tool != "" && origin != ""
}

func normalizedBrowserOrigin(value string) (*url.URL, string, bool) {
	parsed, err := url.Parse(value)
	if err != nil || (parsed.Scheme != "http" && parsed.Scheme != "https") || parsed.Hostname() == "" || parsed.User != nil {
		return nil, "", false
	}
	host := strings.ToLower(strings.TrimSuffix(parsed.Hostname(), "."))
	port := parsed.Port()
	if (parsed.Scheme == "http" && port == "80") || (parsed.Scheme == "https" && port == "443") {
		port = ""
	}
	hostPort := host
	if port != "" {
		hostPort = net.JoinHostPort(host, port)
	}
	normalized := parsed.Scheme + "://" + hostPort
	parsed, _ = url.Parse(normalized)
	return parsed, normalized, true
}

func hostMatches(host, pattern string) bool {
	host = strings.ToLower(strings.TrimSuffix(host, "."))
	pattern = strings.ToLower(strings.TrimSuffix(strings.TrimSpace(pattern), "."))
	if strings.HasPrefix(pattern, "*.") {
		suffix := strings.TrimPrefix(pattern, "*.")
		return suffix != "" && host != suffix && strings.HasSuffix(host, "."+suffix)
	}
	return pattern != "" && host == pattern
}

func originMatches(origin, pattern string) bool {
	parsed, _, ok := normalizedBrowserOrigin(origin)
	if !ok {
		return false
	}
	pattern = strings.TrimSpace(pattern)
	if pattern == "" {
		return false
	}
	if !strings.Contains(pattern, "://") {
		hostPattern, portPattern := pattern, ""
		if host, port, err := net.SplitHostPort(pattern); err == nil {
			hostPattern, portPattern = host, port
		} else if i := strings.LastIndex(pattern, ":"); i > 0 && !strings.Contains(pattern[i+1:], ":") {
			if _, err := strconv.Atoi(pattern[i+1:]); err == nil {
				hostPattern, portPattern = pattern[:i], pattern[i+1:]
			}
		}
		if !hostMatches(parsed.Hostname(), hostPattern) {
			return false
		}
		if portPattern == "" {
			return true
		}
		actualPort := parsed.Port()
		if actualPort == "" {
			if parsed.Scheme == "https" {
				actualPort = "443"
			} else {
				actualPort = "80"
			}
		}
		return actualPort == portPattern
	}
	parts, err := url.Parse(pattern)
	if err != nil || parts.Scheme != parsed.Scheme || !hostMatches(parsed.Hostname(), parts.Hostname()) {
		return false
	}
	return parts.Port() == "" || parts.Port() == parsed.Port()
}

func (p browserOriginPolicy) automaticResponse(params map[string]any) (map[string]any, bool) {
	tool, origin, ok := browserOriginRequest(params)
	if !ok || tool != browserOriginTool {
		return nil, false
	}
	if _, _, valid := normalizedBrowserOrigin(origin); !valid {
		return mcpDecision("decline", nil, nil), true
	}
	switch p.mode {
	case "deny":
		return mcpDecision("decline", nil, nil), true
	case "allow_all":
		return mcpDecision("accept", nil, map[string]any{"persist": "session"}), true
	case "allowlist":
		for _, pattern := range p.allowedOrigins {
			if originMatches(origin, pattern) {
				return mcpDecision("accept", nil, map[string]any{"persist": "session"}), true
			}
		}
	}
	return nil, false
}

func mcpDecision(action string, content any, meta any) map[string]any {
	return map[string]any{"action": action, "content": content, "_meta": meta}
}

func mcpAutomaticResponse(raw json.RawMessage) (map[string]any, bool) {
	params := decodeObject(raw)
	if params == nil {
		return nil, false
	}
	return browserOriginPolicyFromEnv().automaticResponse(params)
}

func mcpElicitationQuestions(params map[string]any) []backend.UserInputQuestion {
	tool, origin, browserRequest := browserOriginRequest(params)
	message := firstNonEmpty(codexFirstString(params, "message"), "Allow this external tool request?")
	server := firstNonEmpty(codexFirstString(params, "serverName"), "MCP request")
	if browserRequest {
		if tool != browserOriginTool {
			return []backend.UserInputQuestion{{
				QuestionID: elicitationChoice, Text: message, Header: "Chrome 高風險授權", Type: "single_select",
				Options: []backend.UserInputOption{
					{ID: "deny", Label: "拒絕", Description: origin, Recommended: true},
					{ID: "approve_once", Label: "僅允許這次", Description: origin},
				},
			}}
		}
		return []backend.UserInputQuestion{{
			QuestionID: elicitationChoice, Text: message, Header: "Chrome 網域授權", Type: "single_select",
			Options: []backend.UserInputOption{
				{ID: "approve_once", Label: "允許這次", Description: origin, Recommended: true},
				{ID: "approve_always", Label: "總是允許", Description: origin},
				{ID: "deny", Label: "拒絕", Description: origin},
			},
		}}
	}

	if codexFirstString(params, "mode") == "url" {
		text := strings.TrimSpace(message + "\n" + codexFirstString(params, "url"))
		return []backend.UserInputQuestion{{
			QuestionID: "action", Text: text, Header: server, Type: "single_select",
			Options: []backend.UserInputOption{{ID: "accept", Label: "Open / Continue", Recommended: true}, {ID: "decline", Label: "Decline"}},
		}}
	}

	schema, _ := params["requestedSchema"].(map[string]any)
	properties, _ := schema["properties"].(map[string]any)
	keys := make([]string, 0, len(properties))
	for key := range properties {
		keys = append(keys, key)
	}
	sort.Strings(keys)
	questions := make([]backend.UserInputQuestion, 0, len(keys))
	for _, key := range keys {
		property, _ := properties[key].(map[string]any)
		propertyType := codexFirstString(property, "type")
		q := backend.UserInputQuestion{
			QuestionID: key,
			Text:       firstNonEmpty(codexFirstString(property, "description", "title"), key),
			Header:     firstNonEmpty(codexFirstString(property, "title"), server),
			Type:       "text",
			FreeForm:   true,
		}
		enumSchema := property
		if propertyType == "array" {
			q.Type, q.MultiSelect = "multi_select", true
			if items, ok := property["items"].(map[string]any); ok {
				enumSchema = items
			}
		}
		q.Options = mcpEnumOptions(enumSchema)
		if propertyType == "boolean" && len(q.Options) == 0 {
			q.Options = []backend.UserInputOption{{ID: "true", Label: "是"}, {ID: "false", Label: "否"}}
		}
		if len(q.Options) > 0 {
			q.FreeForm = false
			if !q.MultiSelect {
				q.Type = "single_select"
			}
		}
		questions = append(questions, q)
	}
	if len(questions) == 0 {
		questions = []backend.UserInputQuestion{{QuestionID: elicitationChoice, Text: message, Header: "外部工具授權", Type: "single_select", Options: []backend.UserInputOption{{ID: "deny", Label: "拒絕", Recommended: true}, {ID: "approve_once", Label: "允許這次"}}}}
	}
	return questions
}

func mcpEnumOptions(schema map[string]any) []backend.UserInputOption {
	values, _ := schema["enum"].([]any)
	options := make([]backend.UserInputOption, 0, len(values))
	for _, value := range values {
		id := codexAnyString(value)
		options = append(options, backend.UserInputOption{ID: id, Label: id})
	}
	return options
}

func mcpElicitationResponse(raw json.RawMessage, answers map[string]any, cancelled bool) map[string]any {
	if cancelled {
		return mcpDecision("cancel", nil, nil)
	}
	decision := codexAnyString(answers[elicitationChoice])
	if decision == "" {
		decision = codexAnyString(answers["action"])
	}
	switch decision {
	case "deny", "decline":
		return mcpDecision("decline", nil, nil)
	case "approve_once":
		return mcpDecision("accept", nil, map[string]any{"persist": "session"})
	case "approve_always":
		return mcpDecision("accept", nil, map[string]any{"persist": "always"})
	}

	params := decodeObject(raw)
	schema, _ := params["requestedSchema"].(map[string]any)
	properties, _ := schema["properties"].(map[string]any)
	content := map[string]any{}
	for key, value := range answers {
		property, exists := properties[key].(map[string]any)
		if !exists {
			continue
		}
		content[key] = coerceMcpAnswer(value, codexFirstString(property, "type"))
	}
	return mcpDecision("accept", content, nil)
}

func coerceMcpAnswer(value any, schemaType string) any {
	switch schemaType {
	case "boolean":
		if boolean, ok := value.(bool); ok {
			return boolean
		}
		parsed, err := strconv.ParseBool(strings.ToLower(strings.TrimSpace(codexAnyString(value))))
		if err == nil {
			return parsed
		}
	case "integer":
		if parsed, err := strconv.Atoi(strings.TrimSpace(codexAnyString(value))); err == nil {
			return parsed
		}
	case "number":
		if parsed, err := strconv.ParseFloat(strings.TrimSpace(codexAnyString(value)), 64); err == nil {
			return parsed
		}
	case "array":
		if _, ok := value.([]any); ok {
			return value
		}
		if value == nil || value == "" {
			return []any{}
		}
		return []any{value}
	}
	return value
}

File: mcp_elicitation.py
"""Codex MCP elicitation normalization and browser-origin policy.

The Codex app-server forwards MCP ``elicitation/create`` requests to its
client.  Browser Use relies on that channel for per-origin consent.  This
module keeps the policy and wire-shape conversion independent from the
app-server transport so it can be tested without launching Codex or Chrome.
"""
from __future__ import annotations

import os
import re
import tempfile
from dataclasses import dataclass
from typing import Any
from urllib.parse import urlsplit


_VALID_BROWSER_MODES = frozenset({"ask", "allowlist", "allow_all", "deny"})
_BROWSER_ORIGIN_TOOL = "access_browser_origin"
_DECISION_QUESTION_ID = "decision"
_FALSE_VALUES = frozenset({"0", "false", "no", "off"})


def _mapping(value: Any) -> dict:
    return value if isinstance(value, dict) else {}


def _walk_mappings(value: Any, *, max_depth: int = 8):
    if max_depth < 0:
        return
    if isinstance(value, dict):
        yield value
        for child in value.values():
            yield from _walk_mappings(child, max_depth=max_depth - 1)
    elif isinstance(value, list):
        for child in value:
            yield from _walk_mappings(child, max_depth=max_depth - 1)


def find_value(payload: Any, key: str) -> Any:
    """Find a metadata field even when an MCP adapter nests ``_meta``."""
    for item in _walk_mappings(payload):
        if key in item:
            return item[key]
    return None


def browser_origin_request(params: dict) -> tuple[str, str] | None:
    """Return ``(tool_name, origin)`` for Browser Use approval elicitations."""
    tool_name = find_value(params, "tool_name")
    origin = find_value(params, "origin")
    connector_id = find_value(params, "connector_id")
    if connector_id != "browser-use" or not isinstance(tool_name, str):
        return None
    if not isinstance(origin, str) or not origin.strip():
        return None
    return tool_name.strip(), origin.strip()


def _normalized_origin(value: str) -> str | None:
    try:
        parsed = urlsplit(value)
    except ValueError:
        return None
    if parsed.scheme not in {"http", "https"} or not parsed.hostname:
        return None
    if parsed.username or parsed.password:
        return None
    host = parsed.hostname.lower().rstrip(".")
    try:
        port = parsed.port
    except ValueError:
        return None
    default_port = (parsed.scheme == "http" and port == 80) or (parsed.scheme == "https" and port == 443)
    suffix = "" if port is None or default_port else f":{port}"
    return f"{parsed.scheme}://{host}{suffix}"


def _host_matches(host: str, pattern: str) -> bool:
    candidate = pattern.strip().lower().rstrip(".")
    if not candidate:
        return False
    if candidate.startswith("*."):
        suffix = candidate[2:]
        return bool(suffix) and host != suffix and host.endswith(f".{suffix}")
    return host == candidate


def origin_matches(origin: str, pattern: str) -> bool:
    normalized = _normalized_origin(origin)
    if normalized is None:
        return False
    parsed = urlsplit(normalized)
    candidate = pattern.strip()
    if not candidate:
        return False

    if "://" not in candidate:
        host_pattern, sep, port = candidate.partition(":")
        if not _host_matches(parsed.hostname or "", host_pattern):
            return False
        return not sep or str(parsed.port or (443 if parsed.scheme == "https" else 80)) == port

    pattern_parts = urlsplit(candidate)
    if pattern_parts.scheme not in {"http", "https"} or pattern_parts.scheme != parsed.scheme:
        return False
    if not _host_matches(parsed.hostname or "", pattern_parts.hostname or ""):
        return False
    try:
        pattern_port = pattern_parts.port
    except ValueError:
        return False
    if pattern_port is None:
        return True
    return parsed.port == pattern_port


@dataclass(frozen=True)
class ElicitationDecision:
    action: str
    content: Any = None
    meta: dict | None = None
    reason: str = ""

    def to_rpc_result(self) -> dict:
        return {"action": self.action, "content": self.content, "_meta": self.meta}


@dataclass(frozen=True)
class BrowserOriginPolicy:
    mode: str = "ask"
    allowed_origins: tuple[str, ...] = ()

    @classmethod
    def from_env(cls) -> "BrowserOriginPolicy":
        raw_mode = os.environ.get("BRIDGE_BROWSER_ORIGIN_MODE", "ask").strip().lower()
        mode = raw_mode if raw_mode in _VALID_BROWSER_MODES else "ask"
        raw_origins = os.environ.get("BRIDGE_BROWSER_ALLOWED_ORIGINS", "")
        allowed = tuple(part.strip() for part in raw_origins.split(",") if part.strip())
        return cls(mode=mode, allowed_origins=allowed)

    def decide(self, params: dict) -> ElicitationDecision | None:
        request = browser_origin_request(params)
        if request is None:
            return None
        tool_name, origin = request

        # Only ordinary origin access may be automated. Raw CDP, history,
        # uploads/downloads, credentials and other Browser Use approvals must
        # remain explicit user interactions.
        if tool_name != _BROWSER_ORIGIN_TOOL:
            return None
        if _normalized_origin(origin) is None:
            return ElicitationDecision("decline", reason="invalid browser origin")
        if self.mode == "deny":
            return ElicitationDecision("decline", reason="browser origin policy denies access")
        if self.mode == "allow_all":
            return ElicitationDecision(
                "accept",
                meta={"persist": "session"},
                reason="browser origin policy allows all valid HTTP(S) origins",
            )
        if self.mode == "allowlist" and any(origin_matches(origin, item) for item in self.allowed_origins):
            return ElicitationDecision(
                "accept",
                meta={"persist": "session"},
                reason="browser origin matched allowlist",
            )
        return None


def ensure_browser_elicitation_routing(codex_home: str | None = None) -> bool:
    """Route Browser Use consent to the app-server client, not auto-review."""
    manage = os.environ.get("BRIDGE_BROWSER_MANAGE_AUTO_REVIEW", "1").strip().lower()
    if manage in _FALSE_VALUES:
        return False
    root = codex_home or os.environ.get("CODEX_HOME") or os.path.expanduser("~/.codex")
    path = os.path.join(root, "browser", "config.toml")
    os.makedirs(os.path.dirname(path), exist_ok=True)
    try:
        with open(path, "r", encoding="utf-8") as handle:
            original = handle.read()
    except FileNotFoundError:
        original = ""

    pattern = re.compile(r"(?m)^[ \t]*disable_auto_review[ \t]*=.*$")
    updated = pattern.sub("disable_auto_review = true", original, count=1)
    if updated == original and not pattern.search(original):
        updated = "disable_auto_review = true\n" + original
    if updated == original:
        return False

    fd, temporary = tempfile.mkstemp(prefix=".config.toml.", dir=os.path.dirname(path))
    try:
        with os.fdopen(fd, "w", encoding="utf-8") as handle:
            handle.write(updated)
            handle.flush()
            os.fsync(handle.fileno())
        if os.path.exists(path):
            os.chmod(temporary, os.stat(path).st_mode & 0o777)
        os.replace(temporary, path)
    finally:
        if os.path.exists(temporary):
            os.unlink(temporary)
    return True


def _enum_options(schema: dict) -> list[dict]:
    values = schema.get("enum")
    names = schema.get("enumNames")
    if isinstance(values, list):
        return [
            {
                "id": str(value),
                "label": str(names[index]) if isinstance(names, list) and index < len(names) else str(value),
                "description": "",
                "recommended": value == schema.get("default"),
            }
            for index, value in enumerate(values)
        ]
    for key in ("oneOf", "anyOf"):
        choices = schema.get(key)
        if isinstance(choices, list):
            result: list[dict] = []
            for choice in choices:
                item = _mapping(choice)
                value = item.get("const", item.get("value", item.get("title", "")))
                result.append({
                    "id": str(value),
                    "label": str(item.get("title") or value),
                    "description": str(item.get("description") or ""),
                    "recommended": value == schema.get("default"),
                })
            return result
    return []


def elicitation_questions(params: dict) -> list[dict]:
    """Convert MCP form/openai-form payloads to the bridge question shape."""
    browser_request = browser_origin_request(params)
    if browser_request is not None:
        tool_name, origin = browser_request
        if tool_name != _BROWSER_ORIGIN_TOOL:
            return [{
                "question_id": _DECISION_QUESTION_ID,
                "text": str(params.get("message") or f"Allow this sensitive Chrome action on {origin}?"),
                "header": "Chrome 高風險授權",
                "type": "choice",
                "options": [
                    {"id": "deny", "label": "拒絕", "description": origin, "recommended": True},
                    {"id": "approve_once", "label": "僅允許這次", "description": origin, "recommended": False},
                ],
                "multi_select": False,
                "free_form": False,
            }]
        return [{
            "question_id": _DECISION_QUESTION_ID,
            "text": str(params.get("message") or f"Allow Chrome to access {origin}?"),
            "header": "Chrome 網域授權",
            "type": "choice",
            "options": [
                {"id": "approve_once", "label": "允許這次", "description": origin, "recommended": True},
                {"id": "approve_always", "label": "總是允許", "description": origin, "recommended": False},
                {"id": "deny", "label": "拒絕", "description": origin, "recommended": False},
            ],
            "multi_select": False,
            "free_form": False,
        }]

    schema = _mapping(params.get("requestedSchema"))
    properties = _mapping(schema.get("properties"))
    required = set(schema.get("required") or [])
    questions: list[dict] = []
    for key, raw_property in properties.items():
        prop = _mapping(raw_property)
        prop_type = str(prop.get("type") or "string")
        options = _enum_options(prop)
        multi = prop_type == "array"
        if multi:
            item_schema = _mapping(prop.get("items"))
            options = _enum_options(item_schema) or _enum_options(prop)
        if prop_type == "boolean" and not options:
            options = [
                {"id": "true", "label": "是", "description": "", "recommended": prop.get("default") is True},
                {"id": "false", "label": "否", "description": "", "recommended": prop.get("default") is False},
            ]
        questions.append({
            "question_id": str(key),
            "text": str(prop.get("description") or prop.get("title") or key),
            "header": str(prop.get("title") or ""),
            "type": "multi_choice" if multi else ("choice" if options else "question"),
            "options": options,
            "multi_select": multi,
            "free_form": not options,
            "required": key in required,
            "schema_type": prop_type,
        })
    if questions:
        return questions
    return [{
        "question_id": _DECISION_QUESTION_ID,
        "text": str(params.get("message") or "Allow this external tool request?"),
        "header": "外部工具授權",
        "type": "choice",
        "options": [
            {"id": "deny", "label": "拒絕", "description": "", "recommended": True},
            {"id": "approve_once", "label": "允許這次", "description": "", "recommended": False},
        ],
        "multi_select": False,
        "free_form": False,
    }]


def _coerce_answer(value: Any, schema: dict) -> Any:
    schema_type = schema.get("type")
    if schema_type == "boolean":
        if isinstance(value, bool):
            return value
        return str(value).strip().lower() in {"1", "true", "yes", "on"}
    if schema_type == "integer":
        try:
            return int(value)
        except (TypeError, ValueError):
            return value
    if schema_type == "number":
        try:
            return float(value)
        except (TypeError, ValueError):
            return value
    if schema_type == "array":
        return value if isinstance(value, list) else ([] if value in (None, "") else [value])
    return value


def elicitation_response(params: dict, response: dict) -> dict:
    """Build a Codex ``McpServerElicitationRequestResponse``."""
    if response.get("cancelled") or response.get("canceled"):
        return {"action": "cancel", "content": None, "_meta": None}
    answers = response.get("answers") if isinstance(response.get("answers"), dict) else {}
    decision = answers.get(_DECISION_QUESTION_ID)
    if decision == "deny":
        return {"action": "decline", "content": None, "_meta": None}
    if decision in {"approve_once", "approve_always"}:
        persist = "always" if decision == "approve_always" else "session"
        return {"action": "accept", "content": None, "_meta": {"persist": persist}}

    schema = _mapping(params.get("requestedSchema"))
    properties = _mapping(schema.get("properties"))
    content = {
        key: _coerce_answer(value, _mapping(properties.get(key)))
        for key, value in answers.items()
        if key in properties
    }
    return {"action": "accept", "content": content, "_meta": None}
